fl4health/server/tabular_feature_alignment_server.py

```diff
@@ -0,0 +1,143 @@
+from fl4health.strategies.basic_fedavg import BasicFedAvg
+
+
+class TabularFeatureAlignmentServer(FlServer):
+    """
+    This server is used when the clients all have tabular data that needs to be
+    aligned.
+
+    tab_features_info: the information that is required for aligning client features.
+    If it is not specified, then the server will randomly poll a client and gather
+    this information from its data source.
+    """
+
+    def __init__(
+        self,
+        client_manager: ClientManager,
+        config: Config,
+        initialize_parameters: Callable,
+        strategy: BasicFedAvg,
+        wandb_reporter: Optional[ServerWandBReporter] = None,
+        checkpointer: Optional[TorchCheckpointer] = None,
+        tab_features_info: Optional[TabularFeaturesInfoEncoder] = None,
+    ) -> None:
+        assert isinstance(strategy, BasicFedAvg)
+        if strategy.on_fit_config_fn is not None:
+            warnings.warn("strategy.on_fit_config_fn will be overwritten.")
+        if strategy.initial_parameters is not None:
+            warnings.warn("strategy.initial_parameters will be overwritten.")
+
+        super().__init__(client_manager, strategy, wandb_reporter, checkpointer)
+        # The server performs one or two rounds of polls before the normal federated training.
+        # The first one gathers feature information if the server does not already have it,
+        # and the second one gathers the input/output dimensions of the model.
+        self.initial_polls_complete = False
+        self.tab_features_info = tab_features_info
+        self.config = config
+        self.initialize_parameters = initialize_parameters
+        self.format_info_gathered = False
+        # casting self.strategy to BasicFedAvg so its on_fit_config_fn can be specified.
+        self.strategy: BasicFedAvg
+        self.strategy.on_fit_config_fn = partial(fit_config, self.config, self.format_info_gathered)
+
+    def fit(self, num_rounds: int, timeout: Optional[float]) -> History:
+        """Run federated averaging for a number of rounds."""
+
+        assert isinstance(self.strategy, BasicFedAvg)
+
+        # Before the normal fitting round begins, the server provides all clients
+        # the feature information needed to perform feature alignment. Then the server
+        # gathers information from the clients that is necessary for initializing the global model.
+        if not self.initial_polls_complete:
+
+            # If the server does not have the needed feature info a priori,
+            # then it requests such information from the clients before the
+            # very first fitting round.
+            if self.tab_features_info is None:
+                # A random client's feature information is selected as the source of truth for feature alignment.
+                feature_info = self.poll_clients_for_feature_info(timeout)
+
+                rand_idx = random.randint(0, len(feature_info) - 1)
```

Comment:
Rather than generating a random index, I believe you can use something like `random.choose(list, 1)` or close to it to just select an element uniformly from a list.